cluster-autoscaler/core/scale_down.go

```diff
@@ -567,27 +566,26 @@ func (sd *ScaleDown) UpdateUnneededNodes(
 			simulator.FindNodesToRemove(
 				currentNonCandidates[:additionalCandidatesPoolSize],
 				destinations,
-				nil,
+				sd.context.ListerRegistry,
 				sd.context.ClusterSnapshot,
 				sd.context.PredicateChecker,
-				additionalCandidatesCount,
-				true,
 				sd.podLocationHints,
 				sd.usageTracker,
 				timestamp,
 				pdbs)
 		if simulatorErr != nil {
 			return sd.markSimulationError(simulatorErr, timestamp)
 		}
+		additionalNodesToRemove = sd.processors.ScaleDownSetProcessor.GetNodesToRemove(additionalNodesToRemove, additionalCandidatesCount)
```

Comment:
I think that inside  `UpdateUnneededNodes`, `maxCount` was supposed not to actually limit anything, since it's equal or higher than the list of candidates that `FindNodesToRemove` iterates over. So this call wouldn't actually be needed, unless the processor wants to return less nodes than it can.

Which brings me to my second point - by putting the processor call here (and inside `getEmptyNodesToRemoveNoResourceLimits`), we're allowing the processor to arbitrarily limit the number of nodes returned by `FindNodesToRemove` inside `UpdateUnneededNodes`, even if there are fewer than `maxCount` nodes passed to it. It's not clear to me if this won't introduce any unexpected issues e.g. with `TryToScaleDown` later on, since it's so tightly coupled with  `UpdateUnneededNodes`.

Is there a use case for calling the same method in two completely different contexts (looking for unneeded nodes and actually choosing nodes to scale down)? The method wouldn't be able to distinguish between the two. If there is a use case, I'd at least introduce two different methods to the processor. But I still think it'd require a bit more thought about how `UpdateUnneededNodes` interacts with `TryToScaleDown`.

Reply:
You are right about those calls inside of `UpdateUnneededNodes` refactored the code to use the `GetNodesToRemove` only just before the scaledowns. So there is no need for the second method.

Reply:
Although, when I looked at the tests it seems this filtering out is needed as it may return more than `additionalCandidatesCount` candidates. This part of the logic aims to limit the number of considered nodes to the `ScaleDownNonEmptyCandidatesCount` limit.
I would say that we should not call the `GetNodesToRemove` here as it picks best candidates and we should take just prefix as in previous version.